File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        freq_map = defaultdict(list)
        for word in strs:
            freq_array = [0] * 26
            for ch in word:
                freq_array[ord(ch)-97] += 1
            freq_map[tuple(freq_array)].append(word)
        return freq_map.values()

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0)
            return new ArrayList();
        HashMap<String, List> map = new HashMap<>(); 
        int[] count = new int[26];
        
        for(String str: strs) {
            Arrays.fill(count, 0);
            
            for(char ch: str.toCharArray()) {
                count[ch-'a']++;
            }
            
            StringBuilder sb = new StringBuilder("");
            
            for(int i = 0; i < 26; ++ i) {
                sb.append("#");
                sb.append(count[i]);
            }
            String key = sb.toString();
            if (!map.containsKey(key)) 
                map.put(key, new ArrayList());
            map.get(key).add(str);
            
        }
        
        return new ArrayList(map.values());
        
    }
}